protocols/abci/addressing-1174.md

```diff
@@ -0,0 +1,230 @@
+
+- consensus _agreement_ is not affected and all processes are still required to agree on the same value..
+- the consensus _validity_ invariant is not affected since we are changing the
+  structure of the `valid(v)` predicate, on which consensus _validity_ directly depends.
+
+However, the new structure of `valid(v)` can affect _termination_ of consensus,
+as some implementations of `ProcessProposal` might reject values
+that CometBFT's internal validity checks would otherwise accept.
+
+This document focuses on how consensus _termination_ is affected
+by the new structure of `valid(v)`, in particular, the different implementations of `ProcessProposal`.
+
+### ABCI 1.0 (and 2.0) Specification
+
+The [ABCI 1.0 specification][abci-spec] imposes a set of new requirements to the application
+so that its implementation of `ProcessProposal` does not compromise _termination_ of consensus,
+given the current CometBFT consensus algorithm
+(called Tendermint, and described in the [arXiv paper][arxiv]).
+We reproduce a slightly modified version of these requirements here.
+This version of the requirements is equivalent to the ones provided in the ABCI 1.0 specification.
+The modifications are needed in order to express other requirements later on in this document.
+
+Let $p$ and $q$ be two correct processes.
+Let $r_p$ be a round of consensus at height $h$ where $p$ is the proposer.
+Let $s_{p,h-1}$ be $p$'s application's state committed for height $h-1$.
+Let $v_p$ be the block that $p$'s CometBFT passes
+on to the application
+via `RequestPrepareProposal` as proposer of round $r_p$, height $h$,
+known as the _raw proposal_.
+Let $u_p$ be the (possibly modified) block that $p$'s application
+returns via `ResponsePrepareProposal` to CometBFT in round $r_p$, height $h$,
+known as the _prepared proposal_.
+
+These are the relevant requirements in the specification of ABCI 1.0 (and 2.0):
+
+* Requirement 3 [`PrepareProposal`, `ProcessProposal`, coherence]: For any two correct processes $p$ and $q$
+  and any round $r_p \geq 0$,
+  if $q$'s CometBFT calls `RequestProcessProposal` on $u_p$,
+  $q$'s application returns Accept in `ResponseProcessProposal`.
+
+* Requirement 4 [`ProcessProposal`, determinism-1]:
+  `ProcessProposal` is a (deterministic) function of the current
+  state and the block being processed.
+  In other words, for any correct process $p$, and any arbitrary block $u$,
+  if $p$'s CometBFT calls `RequestProcessProposal` on $u$ at height $h$,
+  then $p$'s application's acceptance or rejection **exclusively** depends on $u$ and $s_{p,h-1}$.
+
+* Requirement 5 [`ProcessProposal`, determinism-2]:
+  For any two correct processes *p* and *q*, and any arbitrary
+  block $u$,
+  if $p$'s (resp. $q$'s) CometBFT calls `RequestProcessProposal` on *u* at height $h$,
+  then $p$'s application accepts *u* if and only if $q$'s application accepts $u$.
+  Note that this requirement follows from the previous one and consensus _agreement_.
```

Comment:
Requirement 4 is a way to achieve Requirement 5, which is the one that matters.

Reply:
We had a discussion on this when writing the ABCI++ spec. I'm keeping it like this not to diverge from what is written there